### PR TITLE
Imp: some improvements

### LINKED ISSUE #57: Ftr: support provider method name customize
Consider one of following implementation, or all:

> add struct field named method name and DubboName suffix

```go
type UserProvider struct {
	GetUserDubboName struct{} `dubbo:"getUser"`

	user map[string]User
}

func (u *UserProvider) GetUser(ctx context.Context, req []interface{}, rsp *User) error {
	...
}

func (u *UserProvider) Service() string {
	return "com.my.UserProvider"
}

func (u *UserProvider) Version() string {
	return ""
}

```

> add struct `MethodAlias` Method return `map[string]string`  (init needed)

```go
type UserProvider struct {
	user map[string]User
}

func (u *UserProvider) GetUser(ctx context.Context, req []interface{}, rsp *User) error {
	...
}

func (u *UserProvider) Service() string {
	return "com.my.UserProvider"
}

func (u *UserProvider) Version() string {
	return ""
}

func (u UserProvider) MethodAlias() map[string]string {
	return map[string]string{"GetUser":"getUser"}
}
```

### LINKED ISSUE #59: Imp: service provider improvement todo list
service provider improvement todo list:
* allow service provider return nil response
```go
func (p *Provider) GetUser(userID int64) (*User, error) {
	...
	return nil, nil
}
```

* allow service provider none response
```go
func (p *Provider) SetUserInfo(userInfo *User) error {
	...
	return nil
}
```

* allow service provider none request
```go
func (p *Provider) TakeUser() (*User, error) {
	...
	return &User{}
}
```

* host bind ip auto detect when not being configured
* provider config map abstractions
```go
conMap, _ := config.Load()
res, err := conMap["com.ikurento.user.UserProvider"].GetRPCService("com.ikurento.user.UserProvider").(*UserProvider).Echo(context.TODO(), "OK")
if err != nil {
	panic(err)
}
```
to 
```go
providerInstance, _ := config.Build()
res, err := providerInstance.GetRPCService("com.ikurento.user.UserProvider").(*UserProvider).Echo(context.TODO(), "OK")
if err != nil {
	panic(err)
}
```

### LINKED ISSUE #67: Fix: panic when function args list is empty
image
It should not panic when the function args list is empty.

### LINKED ISSUE #79: Improve README to start go-client example
When compiling the source code, remind the user to be careful with the architecture.

```
cd dubbo/go-client
#linux, mac windows represent the os
#release, dev and test represent the environment
sh ./assembly/linux/release.sh
```

For example: 

```
sh ./asssembly/$ARCH/$ENV.sh 

where $ARCH = [linux | mac | windows] and $ENV = [ release | dev | test]

Example: sh ./assembly/linux/release.sh
```

### PR DESCRIPTION
**Pull request by fangyincheng**
**What this PR does**:

    About some improvements.

**Which issue(s) this PR fixes**:

Fixes #79 
Fixes #67 
Fixes #59 
Fixes #57 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

- Improve the way of use in consumer.
- Support none request , none response, nil request, nil response
- Support  lowercase function on the provider
- Fix bug about Port Occupancy when multi-service

### CODE REVIEW

**Comment by gaoxinge** [comment]
LGTM

File: common/proxy/invoker.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	//}
+	//method := svc.Method()[methodName]
+	//if method == nil {
+	//	logger.Errorf("method not found!")
+	//	rpcResult.Err = perrors.New("cannot find method " + methodName + " of svc " + serviceName)
+	//	return rpcResult
+	//}
+	//
+	//returnValues := method.Method().Func.Call(in)
+	return rpcResult
```
Comment (AlexStocks) [inline common/proxy/invoker.go]: 这部分代码功能没有实现，上传上来干嘛呢

File: protocol/dubbo/listener.go [hunk 2]
```diff
@@ -19,6 +19,9 @@ package dubbo
 
 import (
 	"context"
+	"fmt"
+	"github.com/apache/dubbo-go/protocol"
```
Comment (AlexStocks) [inline protocol/dubbo/listener.go]: 新加一个import

File: protocol/jsonrpc/jsonrpc_protocol.go [hunk 3]
```diff
@@ -23,6 +23,8 @@ import (
 	"github.com/apache/dubbo-go/common/logger"
 	"github.com/apache/dubbo-go/config"
 	"github.com/apache/dubbo-go/protocol"
+	"strings"
```
Comment (AlexStocks) [inline protocol/jsonrpc/jsonrpc_protocol.go]: 新 import 语句块

File: common/proxy/invoker.go (hunk 1 above)
Comment (fangyincheng) [reply common/proxy/invoker.go]: Deleted.

File: protocol/dubbo/listener.go (hunk 2 above)
Comment (fangyincheng) [reply protocol/dubbo/listener.go]: Done.

File: protocol/jsonrpc/jsonrpc_protocol.go (hunk 3 above)
Comment (fangyincheng) [reply protocol/jsonrpc/jsonrpc_protocol.go]: Done.

File: config/config_loader.go [hunk 4]
```diff
@@ -236,7 +242,7 @@ func Load() (map[string]*ReferenceConfig, map[string]*ServiceConfig) {
 					(refconfig.Check == nil && consumerConfig.Check == nil) { //default to true
 
 					if refconfig.invoker != nil &&
-						!refconfig.invoker.IsAvailable() {
+						!refconfig.invoker.IsAvailable() { // todo why? it is false when version
```
[the remark this answers is not in the document]
Comment (fangyincheng) [reply config/config_loader.go]: Done.

**Comment by AlexStocks** [comment]
LGTM

**Pull request merged**